torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,795 @@
+        self.batch_dim_size = -1
+
+    def init_batch_dim_size(self, batch_dim_size: int) -> None:
+        """
+        initialize batch dim size base on the first input batch size
+        """
+        if self.batch_dim_size != -1 and self.batch_dim_size != batch_dim_size:
+            raise RuntimeError(
+                f"batch dim size is already initialized! "
+                f"Found new batch size: {batch_dim_size} not "
+                f"matching existing batch dim size: {self.batch_dim_size}!"
+            )
+        self.batch_dim_size = batch_dim_size
+
+    def set_batch_dim(self, node: fx.Node, batch_dim: int) -> None:
+        self.batch_dim_map[node] = batch_dim
+
+    def get_batch_dim(self, node: fx.Node) -> int:
+        if node not in self.batch_dim_map:
+            raise RuntimeError(f"batch dim analysis failed on node: {node}!")
+        return self.batch_dim_map[node]
+
+    def get_batch_dim_shard_spec(
+        self, node: fx.Node, mesh: DeviceMesh, input_full_reduction: bool = False
+    ) -> Tuple[DTensorSpec, bool]:
+        """
+        simple batch dim analysis that analyze the batch dim of the node
+        and return the corresponding shard spec
+        """
+        assert self.batch_dim_size != -1, "batch dim size is not initialized!"
+
+        if node in self.batch_dim_map:
+            node_batch_dim = self.get_batch_dim(node)
+            batch_dim_shard_spec = DTensorSpec(
+                mesh=mesh, placements=[Shard(node_batch_dim)]
+            )
+            return batch_dim_shard_spec, False
+
+        shape = node.meta["val"].shape
+
+        # for reduction op that reduces over the sharded batch dim
+        # we don't generate partial, but rather, we generate shard
+        # This is because the intention of data parallel is to never
+        # do full reduction across batch dimension, it would still
+        # keep the reduction activation sharded.
+        reduction_over_batch = False
+        reduction_ops = [aten.sum.default, aten.mean.default]
+        if node.target in reduction_ops and len(shape) == 0:
+            operand = node.all_input_nodes[0]
+            if operand in self.batch_dim_map:
+                operand_batch_dim = self.get_batch_dim(operand)
+                if operand_batch_dim == 0:
+                    reduction_over_batch = True
+                    self.set_batch_dim(node, operand_batch_dim)
+
+            else:
+                raise RuntimeError(f"batch dim analysis failed on node: {node}!")
+        elif input_full_reduction:
+            # the first consumer node that consumes the full reduction
+            operand = node.all_input_nodes[0]
+            assert (
+                operand in self.batch_dim_map
+            ), "input have full reduction but not in dim map!"
+            self.set_batch_dim(node, self.get_batch_dim(operand))
+        else:
+            for i, dim_size in enumerate(shape):
+                if dim_size == self.batch_dim_size:
```

Comment:
This looks very questionable. At the very least, you should not be willing to mark MULTIPLE dims as the batch dim...